File: src/main/java/src/gestorAplicacion/Calendario/Beca.java
package gestorAplicacion.Calendario;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import gestorAplicacion.personas.Estudiante;
import java.io.Serializable;

public class Beca implements Serializable{

	//Funcionalidad 3: asignación de becas Becas
	private String nombre;
	private tipos tipo;
	private enum tipos {becaInicial, becaNormal, becaAvanzada};
	public static ArrayList<Estudiante> estudiantes;
	protected ArrayList<Estudiante> estudiantesAptosInicial;
	protected ArrayList<Estudiante> estudiantesAptosNormal;
	protected ArrayList<Estudiante> estudiantesAptosAvanzada;
	
	
	public Beca(String nombre) {
		this.nombre = nombre;
		estudiantes = new ArrayList<Estudiante>();
		estudiantesAptosInicial = new ArrayList<Estudiante>();
		estudiantesAptosNormal = new ArrayList<Estudiante>();
		estudiantesAptosAvanzada = new ArrayList<Estudiante>();
	}
	
	
	public ArrayList<Estudiante> getEstudiantes() {
	    return estudiantes;
	}

	public void setEstudiantes(ArrayList<Estudiante> estudiantes) {
	    Beca.estudiantes = estudiantes;
	}
	
	
	public void asignarEstudiantesBeca() {
		ArrayList<Estudiante>estudiantesAptosInicial2 = new ArrayList<Estudiante>();
		ArrayList<Estudiante>estudiantesAptosNormal2 = new ArrayList<Estudiante>();
		ArrayList<Estudiante>estudiantesAptosAvanzada2 = new ArrayList<Estudiante>();
	        
		for (Estudiante estudiante : estudiantes) {
	            if (estudiante.getPorcentajeDeAvance() >= 20 && estudiante.getPorcentajeDeAvance() < 40 &&
	            		estudiante.calcularPromedio() >= 4.5 &&
	            		estudiante.getFueBecado() == false) {
	            	estudiantesAptosInicial2.add(estudiante);
	            	}
	            
	            else if (estudiante.getPorcentajeDeAvance() >= 40 && estudiante.getPorcentajeDeAvance() < 60 &&
	            		estudiante.calcularPromedio() >= 4.0 &&
	            		estudiante.getFueBecado() == false) {
	            	estudiantesAptosNormal2.add(estudiante);
			    	
				}
			    else if(estudiante.getPorcentajeDeAvance() >= 60 && estudiante.getPorcentajeDeAvance() < 100 &&
	            		estudiante.calcularPromedio() >= 3.5 &&
	            		estudiante.getFueBecado() == false){
			    	estudiantesAptosAvanzada2.add(estudiante);
			    }
	        }
	        Collections.sort(estudiantesAptosInicial2, Comparator.comparingDouble(Estudiante::calcularPromedio).reversed());
	        Collections.sort(estudiantesAptosNormal2, Comparator.comparingDouble(Estudiante::calcularPromedio).reversed());
	        Collections.sort(estudiantesAptosAvanzada2, Comparator.comparingDouble(Estudiante::calcularPromedio).reversed());
	        
	        if (estudiantesAptosInicial2.size()>=2) {
	        	for (int i = 0; i<2;i++) {
		        	this.estudiantesAptosInicial.add(estudiantesAptosInicial2.get(i));
	        	}     
	       }
	        else if (estudiantesAptosInicial2.size()==1) {
        		this.estudiantesAptosInicial.add(estudiantesAptosInicial2.get(0));
	        }
	        
	        if (estudiantesAptosNormal2.size()>=2) {
	        	for (int i =0; i<2; i++) {
		        	this.estudiantesAptosNormal.add(estudiantesAptosNormal2.get(i));

	        	}
	        }
	     
	        else if (estudiantesAptosNormal2.size()==1) {
	        	this.estudiantesAptosNormal.add(estudiantesAptosNormal2.get(0));
	        }
	        
	        if (estudiantesAptosAvanzada2.size()>=2) {
	        	for (int i =0; i<2; i++) {
	        		this.estudiantesAptosAvanzada.add(estudiantesAptosAvanzada2.get(i));
	        	}
	        }
	        
	        else if (estudiantesAptosAvanzada2.size()==1) {
	        	this.estudiantesAptosAvanzada.add(estudiantesAptosAvanzada2.get(0));
	        }
		}
}
	

File: src_py/Calendario/Beca.py

from typing import List
from enum import Enum
from src_py.personas.Estudiante import Estudiante

class Beca:
    class Tipos(Enum):
        BECA_INICIAL = 'becaInicial'
        BECA_NORMAL = 'becaNormal'
        BECA_AVANZADA = 'becaAvanzada'

    estudiantes:List['Estudiante']=[]
    
    def __init__(self, nombre):
        self.nombre = nombre
        self.__estudiantes_aptos_inicial = []
        self.__estudiantes_aptos_normal = []
        self.__estudiantes_aptos_avanzada = []


    def getEstudiantes(self):
        return Beca.estudiantes

    def setEstudiantes(self, estudiantes):
        Beca.estudiantes = estudiantes

    def asignar_estudiantes_beca(self):
        estudiantes_aptos_inicial2 = []
        estudiantes_aptos_normal2 = []
        estudiantes_aptos_avanzada2 = []

        for estudiante in Beca.estudiantes:
            if estudiante.porcentaje_de_avance >= 20 and estudiante.porcentaje_de_avance < 40 and estudiante.calcular_promedio() >= 4.5 and not estudiante.getFue_becado():
                estudiantes_aptos_inicial2.append(estudiante)
            elif estudiante.porcentaje_de_avance >= 40 and estudiante.porcentaje_de_avance < 60 and estudiante.calcular_promedio() >= 4.0 and not estudiante.getFue_becado():
                estudiantes_aptos_normal2.append(estudiante)
            elif estudiante.porcentaje_de_avance >= 60 and estudiante.porcentaje_de_avance < 100 and estudiante.calcular_promedio() >= 3.5 and not estudiante.getFue_becado():
                estudiantes_aptos_avanzada2.append(estudiante)

        estudiantes_aptos_inicial2.sort(key=lambda x: x.calcular_promedio(), reverse=True)
        estudiantes_aptos_normal2.sort(key=lambda x: x.calcular_promedio(), reverse=True)
        estudiantes_aptos_avanzada2.sort(key=lambda x: x.calcular_promedio(), reverse=True)

        if len(estudiantes_aptos_inicial2) >= 2:
            for i in range(2):
                self.__estudiantes_aptos_inicial.append(estudiantes_aptos_inicial2[i])
        elif len(estudiantes_aptos_inicial2) == 1:
            self.__estudiantes_aptos_inicial.append(estudiantes_aptos_inicial2[0])

        if len(estudiantes_aptos_normal2) >= 2:
            for i in range(2):
                self.__estudiantes_aptos_normal.append(estudiantes_aptos_normal2[i])
        elif len(estudiantes_aptos_normal2) == 1:
            self.__estudiantes_aptos_normal.append(estudiantes_aptos_normal2[0])

        if len(estudiantes_aptos_avanzada2) >= 2:
            for i in range(2):
                self.__estudiantes_aptos_avanzada.append(estudiantes_aptos_avanzada2[i])
        elif len(estudiantes_aptos_avanzada2) == 1:
            self.__estudiantes_aptos_avanzada.append(estudiantes_aptos_avanzada2[0])

    def getEstudiantes_aptos_inicial(self):
        return self.__estudiantes_aptos_inicial

    def getEstudiantes_aptos_normal(self):
        return self.__estudiantes_aptos_normal

    def getEstudiantes_aptos_avanzada(self):
        return self.__estudiantes_aptos_avanzada
